crates/prover-services/src/parallel/new.rs

```diff
@@ -0,0 +1,329 @@
+        zk_storage: Stf::PreState,
+        thread_pool_size: usize,
+        ledger_db: LedgerDB,
+    ) -> anyhow::Result<Self> {
+        match proof_mode {
+            ProofGenMode::Skip => {
+                tracing::info!("Prover is configured to skip proving");
+            }
+            ProofGenMode::Simulate(_) => {
+                tracing::info!("Prover is configured to simulate proving");
+            }
+            ProofGenMode::Execute => {
+                tracing::info!("Prover is configured to execute proving");
+            }
+            ProofGenMode::Prove => {
+                tracing::info!("Prover is configured to prove");
+            }
+        };
+
+        let thread_pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(thread_pool_size)
+            .build()
+            .expect("Thread pool must be built");
+
+        Ok(Self {
+            thread_pool,
+            proof_mode: Arc::new(Mutex::new(proof_mode)),
+            da_service,
+            vm,
+            zk_storage,
+            ledger_db,
+            current_da_hash: None,
+            proof_queue: vec![],
+        })
+    }
+
+    /// Creates a new `ParallelProverService` with the default number of thread pool size.
+    /// Default thread pool size is num_cpus - 1.
+    pub fn new_with_default_workers(
+        da_service: Da,
+        vm: Vm,
+        proof_mode: ProofGenMode<Da, Vm, Stf>,
+        zk_storage: Stf::PreState,
+        ledger_db: LedgerDB,
+    ) -> anyhow::Result<Self> {
+        let num_cpus = num_cpus::get();
+        assert!(
+            num_cpus > 1,
+            "Parallel prover service requires at least 2 available cores to run smoothly"
+        );
+
+        Self::new(
+            da_service,
+            vm,
+            proof_mode,
+            zk_storage,
+            num_cpus - 1,
+            ledger_db,
+        )
+    }
+
+    pub fn set_current_da_hash(&mut self, da_hash: <Da::Spec as DaSpec>::SlotHash) {
+        assert!(
+            self.current_da_hash.is_none(),
+            "Da hash to prove should never be set twice"
+        );
+        self.current_da_hash = Some(da_hash);
+    }
+
+    pub fn add_proof_data(&mut self, proof_data: ProofData) {
+        assert!(
+            self.current_da_hash.is_some(),
+            "Add proof data should never be called before setting da hash"
+        );
+        self.proof_queue.push(proof_data);
+    }
+
+    pub async fn prove_and_submit(
+        &mut self,
+    ) -> anyhow::Result<Vec<<Da as DaService>::TransactionId>> {
+        if let ProofGenMode::Skip = *self.proof_mode.lock() {
+            tracing::debug!(
+                "Skipped proving {} proofs in block {:?}",
+                self.proof_queue.len(),
+                self.current_da_hash
+            );
+
+            self.current_da_hash = None;
+            self.proof_queue.clear();
+
+            return Ok(vec![]);
+        }
+
+        assert!(
```

Comment:
Assert = panic

Reply:
I think it is fine to panic, if it is a bug. Because there is no recovery back from a bug. And I think it makes it easier to actually figure the bug out instead of proving being finished immediately and trying to debug why is that.